Update NXP LPC DMA driver for complete_callback_en flag

mmahadevan108 opened the pull request:
Add code to recognize the complete_callback flag and issue the callback accordingly.

mmahadevan108 committed:
drivers: dma_lpc: Process complete_callback flag

Add code to recognize the complete_callback flag
and issue the callback accordingly.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: spi: Set the complete_callback flag

The MCUX LPC driver now recognizes the complete_callback
flag. We need to set this flag so we receive a callback
after every block.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: i2s: Set DMA complete_callback flag

The MCUX LPC DMA driver now recognizes the compelete_callback
flag. Set this flag so we receive an interrupt after completion
of every block.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -543,7 +586,16 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
-			if (dma_mcux_lpc_queue_descriptors(data, block_config, src_inc, dst_inc)) {
+			if (block_config->next_block == NULL) {
+				/* This is the last block. If complete callback is called only
+				 * on the last block then enable it for the last block
```
i am confused by the wording of this comment

mmahadevan108 replied:
I have updated the comment.

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -461,6 +478,11 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
 
+	if (config->complete_callback_en) {
+		complete_callback = true;
+	}
```
just curious is it possible to do `complete_callback = (bool) config->complete_callback_en` or something like that?

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -477,13 +499,24 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+				/* Disable interrupt, only enable at the end of transfer
+				 * Reload enable.
+				 */
+				xfer_config = DMA_CHANNEL_XFER(1UL, 0UL, 0UL, 0UL,
```
instead of if/else, can it just be `DMA_CHANNEL_XFER(1, 0, complete_callback, ...)` ?

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -527,7 +560,17 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+				 */
+				if (!complete_callback) {
+					complete_callback = true;
+				}
```
maybe the `if(!complete_callback)` is not needed here since in this case we are always setting it to true it seems like

mmahadevan108 replied:
We still need this piece of code as the earlier set to true was for the case when we queue only 1 block.

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -238,12 +243,23 @@ static int dma_mcux_lpc_queue_descriptors(struct channel_data *data,
+			/* Disable interrupt, we don't want an interrupt after this block
+			 * enable reload for the descriptor
+			 */
+			xfer_config = DMA_CHANNEL_XFER(1UL, 0UL, 0U, 0U,
```
similar to other comment, can it just be `DMA_CHANNEL_XFER(1, 0, callback_en, ...)`

mmahadevan108 replied:
Thanks, makes sense. I have made the change.

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -188,7 +189,11 @@ static int dma_mcux_lpc_queue_descriptors(struct channel_data *data,
+				enable_interrupt = 1;
+			} else {
+				enable_interrupt = 0;
+			}
```
`enable_interrupt = callback_en` ?

mmahadevan108 replied:
Thanks, I have made this change.

yvanderv reviewed (commented):
A couple of optimizations like Declan is suggesting. Otherwise, it looks good

decsny commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -544,7 +556,14 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
+			if (block_config->next_block == NULL) {
+				/* This is the last block. Enable callback if not enabled. */
+				if (!complete_callback) {
+					complete_callback = true;
```
dont think the `if` is needed here

mmahadevan108 replied:
This is still needed. The previous time this was set was for the case where there is only 1 DMA block.

decsny replied:
for some reason, I did not see the `!`, and thought it was redundant

decsny replied:
@mmahadevan108 if you are setting it to true unconditionally why does there need to be a check of if it is false

mmahadevan108 replied:
This is done only for the case when we have not additional blocks in the queue. Hence it does not apply for the code further.

mmahadevan108 replied:
Thanks, I have made this change.

mmahadevan108 commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -478,9 +485,8 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
-			 */
-			xfer_config = DMA_CHANNEL_XFER(1UL, 0UL, 1UL, 0UL,
+			/* Enable interrupt if user requested and reload for the descriptor */
+			xfer_config = DMA_CHANNEL_XFER(1UL, 0UL, complete_callback, 0UL,
```
that sounds like a good idea. I will look into updating the PR with this implementation.

mmahadevan108 replied:
Implementation updated. I have added this change as a separate commit.

mmahadevan108 committed:
drivers: dma: Return DMA_STATUS_BLOCK or DMA_STATUS_COMPLETE

Use intA and intB fields of the DMA descriptor to decide when
the interrupt is per block versus when the transfer is complete.
This allows us to return the correct flag to the user.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
@yvanderv @decsny @danieldegrasse can you revisit this PR?

mmahadevan108 commented:
> with this change is it possible for the i2s_api test to pass on mimxrt595_evk_cm33 ?
The I2S driver is setting this bit and expects a callback on each block. This change avoids calling the callback on each block if the user configures in this manner.
